### DISCUSSION TITLE
new ServiceClient "hanging"

### DISCUSSION THREAD

**Discussion by mjfpalmer** [microsoft/PowerPlatform-DataverseServiceClient#212] [Q&A]
I'm creating a .Net 6.0 Blazor Server project which references version  0.5.10 of the DataverseServiceClient and injects a Singleton wrapper class containing an instance of a ServiceClient for use in Controllers and Razor pages/components through DI.

If I instantiate an instance of ServiceClient in Program.cs _before_ the application starts running, it connects fine, IsReady is true and the ServiceClient can retrieve/create/etc. This confirms that my CRM connection string is correct. The app connects to CRM using the ClientSecret AuthType.

If I attempt to use the class after the app has started running, when it attempts to instantiate a new ServiceClient using the constructor below, then the application hangs on this line forever without ever throwing an exception or running past the instantiation.
`this.ServiceClient = new ServiceClient(organisationURL, clientId, clientSecret, true);`

The only information I have about what is going on is via app insights, where I can see the events logged below (sensitive information replaced by XXX).

**Event 1**
```
Application Insights Telemetry: {"name":"AppDependencies","time":"2021-11-26T20:20:11.0378650Z","iKey":"XXX","tags":{"ai.application.ver":"1.0.0.0","ai.cloud.roleInstance":"XXX","ai.operation.id":"XXX","ai.operation.parentId":"XXX","ai.operation.name":"GET /_Host","ai.location.ip":"::1","ai.internal.sdkVersion":"rdddsc:2.19.0-227","ai.internal.nodeName":"XXX"},"data":{"baseType":"RemoteDependencyData","baseData":{"ver":2,"name":"GET /XRMServices/2011/Organization.svc","id":"XXX","data":"https://XXX.crm4.dynamics.com:443/XRMServices/2011/Organization.svc/web?SDKClientVersion=9.2.46.6061","duration":"00:00:00.8216364","resultCode":"401","success":false,"type":"WCF Service","target":"XXX.crm4.dynamics.com","properties":{"DeveloperMode":"true","_MS.ProcessedByMetricExtractors":"(Name:'Dependencies', Ver:'1.1')","AspNetCoreEnvironment":"Development"}}}}
```

**Event 2**
```
Application Insights Telemetry: {"name":"AppDependencies","time":"2021-11-26T20:20:11.9733863Z","iKey":"XXX","tags":{"ai.application.ver":"1.0.0.0","ai.cloud.roleInstance":"XXX","ai.operation.id":"XXX","ai.operation.parentId":"XXX","ai.operation.name":"GET /_Host","ai.location.ip":"::1","ai.internal.sdkVersion":"rdddsc:2.19.0-227","ai.internal.nodeName":"XXX"},"data":{"baseType":"RemoteDependencyData","baseData":{"ver":2,"name":"GET /common/discovery/instance","id":"XXX","data":"https://login.microsoftonline.com:443/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2FXXX%2Foauth2%2Fv2.0%2Fauthorize","duration":"00:00:00.7891091","resultCode":"200","success":true,"type":"Http","target":"login.microsoftonline.com","properties":{"DeveloperMode":"true","_MS.ProcessedByMetricExtractors":"(Name:'Dependencies', Ver:'1.1')","AspNetCoreEnvironment":"Development"}}}}
```

I realise this is a fairly specific situation, but I'm really looking for any guidance on the questions below:

1. Is the initial HTTP 401 response expected in **Event 1** expected before **Event 2**?
2. Can you shed any light on why the ServiceClient instantiation "hangs".

**Comment by mjfpalmer** [comment]
Ah, I just found https://github.com/microsoft/PowerPlatform-DataverseServiceClient/issues/113 which seems to deal with the same issue. I will report back soon as to whether the suggested fixes there resolve my issue.

That issue is closed, but I've got evidence that this issue still exists in 0.5.10.

**Reply by MattB-msft** [reply]
There is some wonky (technical term here) behavior with Auth under some circumstances that is inside the MSAL libs.. 
We have been working to fix thread blocking on our side and we believe we got them all on our side. 

if your still seeing it, please open an issue with a repro of what your doing and we will take a look.

**Reply by mjfpalmer** [reply]
Thanks @MattB-msft.

The fix described in #113  has worked to get me around the issue, but without it, I am definitely still seeing the "hang" behaviour.

I'll create an issue...
